### PR TITLE
Unify CosSinRoPE and ComplexRoPE YaRN computation

### PR DESCRIPTION
**Pull request by sdmyzlp**
### Background

`ComplexRoPE` and `CosSinRoPE` are the two RoPE variants in torchtitan: `ComplexRoPE` uses an interleaved complex cache (DeepSeek V3, LLaMA 3, Qwen3.5); `CosSinRoPE` uses a concatenated cos/sin cache with `rotate_half` application (GPT-OSS, Qwen3). Both implement YaRN ("NTK-by-parts") for context-length extension, but they drifted apart in following ways:

1. **`beta_fast`/`beta_slow` assignment** — `CosSinRoPE` mapped `low ← beta_slow`, `high ← beta_fast`, **the inverse of** the YaRN paper convention and of `ComplexRoPE`. GPT-OSS's config compensated with swapped values, hiding the inconsistency from users.

2. **Embedded rope `mscale`** — `CosSinRoPE` multiplied the YaRN attention-magnitude factor (`0.1·log(factor)+1`) into its cos/sin cache, coupling the rope to the attention. GPT-OSS needed this; other `rotate_half` consumers may not. The rope should be a pure rotation, with mscale applied where each model's attention expects it.

This PR makes the two classes share one YaRN implementation, agreeing exactly when both enter their respective YaRN (or non-YaRN) branches.

### Commits

**1. Align `CosSinRoPE` `beta_fast`/`beta_slow` to the YaRN paper/HF convention**
Swap `low ← beta_slow` to `low ← beta_fast` (extrapolation), `high ← beta_fast` to `high ← beta_slow` (interpolation). Un-swap GPT-OSS config to the official `beta_fast=32.0, beta_slow=1.0`. Numerically neutral (the two un-swaps cancel).

**2. Extract a shared `_yarn_inv_freq` helper + `truncate` flag**
A single source-of-truth function called by both classes. Add a `truncate` config flag (floor/ceil for DeepSeek; fractional for GPT-OSS, which uses `truncate=False`). Always clamps to `[0, dim-1]`.

**3. Make `CosSinRoPE` a pure rotation; let GPT-OSS handle mscale in its attention**
Remove `mscale` from `CosSinRoPE._precompute_cache` (and the unused `mscale` field from `RoPE.Config`). GPT-OSS folds `mscale²` into its `softmax_scale` — mathematically equivalent. The rope is now modular: each model's attention owns its own mscale.

### References

- YaRN paper: [arXiv:2309.00071](https://arxiv.org/abs/2309.00071) (§3.2 NTK-by-parts, §3.3 attention temperature)
- DeepSeek-V3.2: [`precompute_freqs_cis`](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/blob/c69397e/inference/model.py#L324-L402) — `low = floor(fcd(beta_fast))`, `high = ceil(fcd(beta_slow))`
- GPT-OSS official: [`ntk_alpha=1, ntk_beta=32`](https://github.com/openai/gpt-oss/blob/d21f34e/gpt_oss/torch/model.py#L28-L29); [`low ← ntk_beta` in rope](https://github.com/openai/gpt-oss/blob/d21f34e/gpt_oss/torch/model.py#L85-L131)

### CODE REVIEW

**Comment by sdmyzlp** [comment]
@shuhuayu Could you please take a look at this? Thanks.

File: torchtitan/models/common/rope.py [hunk 1]
```diff
@@ -62,7 +61,9 @@ class Config(Module.Config):
         beta_fast: float = 32.0
         beta_slow: float = 1.0
         original_seq_len: int = 4096
-        mscale: float = 0.0
+        # Floor/ceil the YaRN correction range (DeepSeek style). gpt-oss uses
+        # fractional cutoffs (``truncate=False``).
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: nit: remove

File: torchtitan/models/common/rope.py [hunk 2]
```diff
@@ -133,6 +134,50 @@ def _init_self_buffers(self, *, buffer_device: torch.device | None = None) -> No
             self.cache = self._precompute_cache()
 
 
+def _yarn_inv_freq(
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: nit: move it up, after _maybe_check_max_pos.

File: torchtitan/models/common/rope.py [hunk 3]
```diff
@@ -133,6 +134,50 @@ def _init_self_buffers(self, *, buffer_device: torch.device | None = None) -> No
...
+        high = math.ceil(high)
+    low, high = max(low, 0), min(high, dim - 1)
+    assert 0 < low < high < dim - 1, (
+        f"Invalid YaRN params: 0 < {low} < {high} < {dim - 1}"
+    )
+
+    ramp = ((torch.arange(dim // 2, dtype=torch.float32) - low) / (high - low)).clamp(
+        0, 1
+    )
+    smooth = 1 - ramp
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: can you drop `smooth` and use `ramp` in the `return` line directly?

File: torchtitan/models/common/rope.py [hunk 4]
```diff
@@ -269,49 +282,33 @@ def _precompute_cache(self) -> torch.Tensor:
...
-                * math.log(cfg.original_seq_len / (cfg.beta_slow * 2 * math.pi))
-                / math.log(base)
-            )
-            high = (
-                d_half
-                * math.log(cfg.original_seq_len / (cfg.beta_fast * 2 * math.pi))
-                / math.log(base)
+            # Shared YaRN math (pure rotation). The YaRN attention "mscale" is
+            # applied by the model in its softmax scale, not baked into the rope
+            # cos/sin -- see _yarn_inv_freq.
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: nit: remove

File: torchtitan/models/common/rope.py [hunk 5]
```diff
@@ -173,48 +218,16 @@ def _precompute_cache(self) -> torch.Tensor:
...
-                if min_val == max_val:
-                    max_val += 0.001
-                linear_func = (torch.arange(dim, dtype=torch.float32) - min_val) / (
-                    max_val - min_val
-                )
-                return torch.clamp(linear_func, 0, 1)
-
-            low, high = find_correction_range(
-                beta_fast, beta_slow, dim, base, original_seq_len
+            # YaRN (DeepSeek V3 style); see _yarn_inv_freq for the shared math.
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: nit: 
```suggestion
            # YaRN (DeepSeek V3 style)
```

File: torchtitan/models/common/rope.py [hunk 6]
```diff
@@ -269,49 +282,33 @@ def _precompute_cache(self) -> torch.Tensor:
...
         freqs = torch.outer(t, inv_freq).float()
-        theta = torch.cat([freqs, freqs], dim=-1)
-
-        cos = theta.cos() * mscale
-        sin = theta.sin() * mscale
-        return torch.cat([cos, sin], dim=-1)
+        cis = torch.polar(torch.ones_like(freqs), freqs)
+        cos_dup = torch.cat([cis.real, cis.real], dim=-1)
+        sin_dup = torch.cat([cis.imag, cis.imag], dim=-1)
+        return torch.cat([cos_dup, sin_dup], dim=-1)
```
Comment (shuhuayu) [inline torchtitan/models/common/rope.py]: i do not see why we need to change to this way to compute cos sin cache, maybe revert to the previous math. in the long term, it is doable to unify both ropes and users can config `interleaved` or `split_half`, and store the cache in cos/sin or complex.

cc: @tianyu-l

File: torchtitan/models/gpt_oss/model.py [hunk 7]
```diff
@@ -104,6 +72,13 @@ def __init__(self, config: Config):
 
         # Standard attention softmax scale (1/sqrt(head_dim))
         self.softmax_scale = 1.0 / math.sqrt(self.head_dim)
+        # YaRN attention "mscale": the rope is now a pure rotation (this factor
+        # used to be baked into the rope cos/sin). Apply the magnitude factor
+        # (mscale**2 -- equivalent to scaling q and k each by mscale) here in the
+        # softmax scale instead. See _yarn_inv_freq in models/common/rope.py.
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model.py]: nit: remove

File: torchtitan/models/gpt_oss/model.py [hunk 8]
```diff
@@ -104,6 +72,13 @@ def __init__(self, config: Config):
...
+        if config.rope.scaling == "yarn" and config.rope.rope_factor > 1.0:
+            mscale = 0.1 * math.log(config.rope.rope_factor) + 1.0
```
Comment (shuhuayu) [inline torchtitan/models/gpt_oss/model.py:108]: nit: maybe add one line comment saying: Merge YaRN attention mscale into softmax_scale (mscale**2 == scaling q and k each by mscale)

**Review by shuhuayu — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/rope.py (hunk 6 above)
Comment (shuhuayu) [reply torchtitan/models/common/rope.py]: thx for the update. can you revert to the original `.cos()`, `.sin()` approach to compute cache, `torch.polar` brings some small numerical differences here. and using `.cos()` `.sin()` is more faithful to its name `CosSinRoPE`.

**Review by shuhuayu — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/rope.py (hunk 6 above)
Comment (sdmyzlp) [reply torchtitan/models/common/rope.py]: Thanks for the review! I've addressed the other comments already.

Could we revisit this one briefly? This is for DeepSeek-V3.2: the Indexer uses [non-interleaved](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/blob/main/inference/model.py#L469) (rotate_half) while the main model uses interleaved, so `CosSinRoPE` fits the Indexer better than adding `interleaved` to `ComplexRoPE`. But the ~1 ULP mismatch between `torch.cos/sin` and `torch.polar` made reuse messy.

Would love to hear your thoughts on this!

Comment (sdmyzlp) [reply torchtitan/models/common/rope.py]: > `torch.polar` brings some small numerical differences here.

Yes this is expected. My preference for polar is really about keeping DSv3.2 internally consistent, but I don't want to break existing cases either. Will revert that last commit if this does hurt.

Comment (shuhuayu) [reply torchtitan/models/common/rope.py]: Thanks for the reference. I am reading the [rope for indexer](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/blob/main/inference/model.py#L464), which uses non-interleaved rope as you mentioned. 

1. Per my reading, `torch.polar`, i.e., ComplexRoPE, is used for [both attention and indexer](https://huggingface.co/deepseek-ai/DeepSeek-V3.2/blob/main/inference/model.py#L401), so to match it bit-wisely, we should use `ComplexRoPE` with `interleaved` option.

2. If we use `CosSinRoPE` with `torch.ploar`, this is not bit-wise identical to `view_as_complex`. PyTorch's complex multiply rounds differently from explicit real multiply-add (`CosSinRoPE` with `torch.ploar` as used in current pr). You may write a small test to verify. I did a small one offline. So the current form does not serve the purpose. 

3. The current main `CosSinRoPE` is used in qwen3 and qwen3.5, which does not use `torch.polar` per their hf implementations so we have reasons to keep them.

4. I think to support deepseek v3.2, a cleaner way is to add an `interleaved` option to `ComplexRoPE(interleaved=False)`, which matches the reference bitwise. Since it's independent of this pr, would you mind doing it as a separate follow-up pr?

Comment (sdmyzlp) [reply torchtitan/models/common/rope.py]: I have dropped that commit, thanks for the clarification!

Comment (shuhuayu) [reply torchtitan/models/common/rope.py]: cool, please fix the lint.

**Review by shuhuayu — APPROVED** [review]
LGTM!

File: torchtitan/models/gpt_oss/__init__.py [hunk 9]
```diff
@@ -283,8 +277,9 @@ def _debugmodel(
                 theta=150000.0,
                 scaling="yarn",
                 rope_factor=32,
-                beta_slow=32.0,
-                beta_fast=1.0,
+                beta_fast=32.0,
+                beta_slow=1.0,
+                truncate=False,
```
Comment (tianyu-l) [inline torchtitan/models/gpt_oss/__init__.py:288]: This makes it different from default value.

Is this fixing a bug, or just changing the way it's written (by updating `softmax_scale` in gpt-oss model.py).

**Review by tianyu-l — COMMENTED** [review]
numerics CI seems broken https://github.com/pytorch/torchtitan/actions/runs/28502028306/job/84481564730?pr=3787

@shuhuayu could you make sure the main CI is healthy, and then ask for a rebase and verify this PR doesn't change numerics, and then merge?

File: torchtitan/models/gpt_oss/__init__.py (hunk 9 above)
Comment (shuhuayu) [reply torchtitan/models/gpt_oss/__init__.py:288]: this is a remap for `beta_fast` and `beta_slow`, so it should be fine. currently, 

```
  ComplexRoPE (rope.py:200-215) — uses the YaRN-paper convention:
  low  = math.floor(find_correction_dim(low_rot,  ...))   # low_rot  = beta_fast
  high = math.ceil (find_correction_dim(high_rot, ...))   # high_rot = beta_slow
  low, high = find_correction_range(beta_fast, beta_slow, ...)   # low <- beta_fast, high <- beta_slow

  CosSinRoPE (rope.py:286-293) — inverted:
  low  = ... math.log(orig / (cfg.beta_slow * 2*pi)) ...   # low  <- beta_slow
  high = ... math.log(orig / (cfg.beta_fast * 2*pi)) ...   # high <- beta_fast
```
and this pr unified it.

**Comment by shuhuayu** [comment]
I think the current main ci failed the numerics test, we can check it first and land this pr later.

**Review by tianyu-l — APPROVED** [review]

**Comment by sdmyzlp** [comment]
Got it, will wait for the main CI issue to be resolved.

**Comment by sdmyzlp** [comment]
Seems the lint/rocm CI is not fully recovered.

**Comment by shuhuayu** [comment]
> Seems the lint/rocm CI is not fully recovered.

yeah, we are trying to fix it.

**Comment by sdmyzlp** [comment]
> yeah, we are trying to fix it.

Hi @shuhuayu — quick question on this PR. Is the main CI healthy now? Would it be possible to re-run the checks and revisit merging? Thanks!
